packages/types/types/src/matrix.ts

```diff
@@ -4,49 +4,70 @@
  */
 
 /** An observable 2D numerically indexed collection. */
-export interface IMatrixProducer<T> {
+export interface IMatrixShapeProducer {
     /**
      * Acquire a reader for this matrix's values and implicitly subscribe the consumer
      * to value change notifications.
-     * 
+     *
      * @param consumer - The consumer to be notified of matrix changes.
      */
-    openMatrix(consumer: IMatrixConsumer<T>): IMatrixReader<T>;
+    openMatrix(consumer: IMatrixShapeConsumer): IMatrixShapeReader;
 
     /**
      * Unsubscribe the consumer from this matrix's change notifications.
-     * 
+     *
      * @param consumer - The consumer to unregister from the matrix.
      */
-    closeMatrix(consumer: IMatrixConsumer<T>): void;
+    closeMatrix(consumer: IMatrixShapeConsumer): void;
+}
+
+/** A consumer of change notifications for a matrix. */
+export interface IMatrixShapeConsumer {
+    /** Notification that rows have been inserted, removed, and/or replaced in the given matrix. */
+    rowsChanged(rowStart: number, removedCount: number, insertedCount: number, producer: IMatrixShapeProducer): void;
+
+    /** Notification that cols have been inserted, removed, and/or replaced in the given matrix. */
+    colsChanged(colStart: number, removedCount: number, insertedCount: number, producer: IMatrixShapeProducer): void;
 }
 
 /** Capability to read cells in a matrix. */
-export interface IMatrixReader<T> {
+export interface IMatrixShapeReader {
     readonly rowCount: number;
     readonly colCount: number;
-    getCell(row: number, col: number): T;
 
     /**
      * A reference to the underlying producer that provides values for this reader,
      * or undefined if the producer is immutable.
      */
-    readonly matrixProducer?: IMatrixProducer<T>;
+    readonly matrixProducer?: IMatrixShapeProducer;
 }
 
 /** Capability to write cells in a matrix. */
-export interface IMatrixWriter<T> {
-    setCell(row: number, col: number, value: T): void;
+export interface IMatrixShapeWriter {
+    spliceRows(rowStart: number, deleteCount: number, insertCount: number): void;
+    spliceCols(colStart: number, deleteCount: number, insertCount: number): void;
 }
 
-/** A consumer of change notifications for a matrix. */
-export interface IMatrixConsumer<T> {
-    /** Notification that rows have been inserted, removed, and/or replaced in the given matrix. */
-    rowsChanged(rowStart: number, removedCount: number, insertedCount: number, producer: IMatrixProducer<T>): void;
+/** An observable 2D numerically indexed collection. */
+export interface IMatrixProducer<T> extends IMatrixShapeProducer {
```

Comment:
Generally, an I*Producer is implicitly an I*ShapeProducer, an I*Reader is an I*ShapeReader, and so on...

The exceptions are:
* I*Writers are not I*ShapeWriters.   (e.g., the capability to set cells in a matrix is separate from the capability to resize a matrix.)
* IProducer is not an IShapeProducer because not all IProducers support reflection.  (At least, currently.)

Reply:
The co and contra variance between reading and writing feels pretty natural.

Value reading implies shape reading; I need to know the shape to read a value.
Shape writing implies value writing; to change the shape I need to create or destroy exiting values.

but,

I can read a shape without reading the values, and I can write values without writing the shape.

Reply:
There's one subtly when shapes are shared.  For example, I create a resizable matrix like this:
```ts
const values: IMatrixWriter & IMatrixShapeWriter = new ResizableMatrix();
```
And someone else creates a matrix that shares the same shape:
```ts
const annotations: IMatrixWriter = new Matrix(/* shapeProducer: */ values);
```
In this situation 'annotations' is an IMatrixWriter, but not an IShapeMatrixWriter.  This means that value producers need to handle key insertion w/o an associated value by producing an empty/default value like 'undefined'.

This does create a "glitch" between key insertion and setting the values:
```ts
values.insertRows(0, 10);    // inserts 10 empty rows
values.setValues(0, 0, /* colCount: */ 3, [ ... ]);  // add associated values 
annotations.setValues(0, 0, /* colCount: */ 3, [ ... ]);  // add associated annotations 
```
IMO, the way to solve this is some form of implicit transaction, such as lazy recalc or deferring/scheduling recalc for the next microtask.

Value reading does imply shape reading, except for "records", which are maps with an immutable shape.  My thinking was that since the shape is immutable, we shouldn't require the implementer to provide 'keys()'/'has()'...  but it's a bit of an experiment.  I can easily see us backtracking on that one.